File: main/src/main/java/com/sedmelluq/discord/lavaplayer/tools/exception/EnvironmentInformation.java
package com.sedmelluq.discord.lavaplayer.tools.exception;

import com.sedmelluq.discord.lavaplayer.tools.PlayerLibrary;

public class EnvironmentInformation extends Exception {
  private static final String[] PROPERTIES = new String[] {
      "os.arch",
      "os.name",
      "os.version",
      "java.vendor",
      "java.version",
      "java.runtime.version",
      "java.vm.version"
  };

  public static final EnvironmentInformation INSTANCE = create();

  private EnvironmentInformation(String message) {
    super(message, null, false, false);
  }

  private static EnvironmentInformation create() {
    DetailMessageBuilder builder = new DetailMessageBuilder();
    builder.appendField("lavaplayer.version", PlayerLibrary.VERSION);

    for (String property : PROPERTIES) {
      builder.appendField(property, System.getProperty(property));
    }

    return new EnvironmentInformation(builder.toString());
  }
}

File: main/src/main/java/lavaplayer/tools/exception/EnvironmentInformation.kt
package lavaplayer.tools.exception

import lavaplayer.tools.PlayerLibrary
import java.lang.Exception

class EnvironmentInformation private constructor(message: String) : Exception(message, null, false, false) {
    companion object {
        private val PROPERTIES = arrayOf(
            "os.arch",
            "os.name",
            "os.version",
            "java.vendor",
            "java.version",
            "java.runtime.version",
            "java.vm.version"
        )

        val INSTANCE = create()

        private fun create(): EnvironmentInformation {
            val builder = DetailMessageBuilder()
            builder.appendField("lavaplayer.version", PlayerLibrary.VERSION)
            for (property in PROPERTIES) {
                builder.appendField(property, System.getProperty(property))
            }

            return EnvironmentInformation(builder.toString())
        }
    }
}
